go/heif/heif.go

```diff
@@ -602,43 +694,259 @@ func (img *Image) GetImage() (image.Image, error) {
+	options := &EncodingOptions{
+		options: C.heif_encoding_options_alloc(),
+	}
+	if options.options == nil {
+		return nil, fmt.Errorf("Could not allocate encoding options")
+	}
+
+	runtime.SetFinalizer(options, freeHeifEncodingOptions)
+	return options, nil
+}
+
+func freeHeifEncodingOptions(options *EncodingOptions) {
+	C.heif_encoding_options_free(options.options)
+	options.options = nil
+	runtime.SetFinalizer(options, nil)
+}
+
+func hasAlpha(i image.Image) bool {
+	rect := i.Bounds()
+	for y := rect.Min.Y; y < rect.Max.Y; y++ {
+		for x := rect.Min.X; x < rect.Max.X; x++ {
+			if _, _, _, c := i.At(x, y).RGBA(); c == 0xffff {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func imageFromNRGBA(i *image.NRGBA) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	var cm Chroma
+	var depth int
+	var stride int
+	if hasAlpha(i) {
+		cm = ChromaInterleavedRGBA
+		depth = 32
+		stride = w * 4
+	} else {
+		cm = ChromaInterleavedRGB
+		depth = 64
+		stride = w * 3
+	}
+
+	out, err := NewImage(w, h, ColorspaceRGB, cm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	p, err := out.NewPlane(ChannelInterleaved, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add plane: %v", err)
+	}
+	p.SetData([]byte(i.Pix), stride)
+
+	return out, nil
+}
+
+func imageFromGray(i *image.Gray) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	out, err := NewImage(w, h, ColorspaceYCbCr, ChromaMonochrome)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	const depth = 8
+	pY, err := out.NewPlane(ChannelY, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Y plane: %v", err)
+	}
+	pY.SetData([]byte(i.Pix), i.Stride)
+
+	return out, nil
+}
+
+func imageFromYCbCr(i *image.YCbCr) (*Image, error) {
+	max := i.Bounds().Max
+	w, h := max.X, max.Y
+
+	out, err := NewImage(w, h, ColorspaceYCbCr, Chroma420)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image: %v", err)
+	}
+
+	const depth = 8
+	pY, err := out.NewPlane(ChannelY, w, h, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Y plane: %v", err)
+	}
+	pY.SetData([]byte(i.Y), i.YStride)
+
+	halfW, halfH := (w+1)/2, (h+1)/2
+	pCb, err := out.NewPlane(ChannelCb, halfW, halfH, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Cb plane: %v", err)
+	}
+	pCb.SetData([]byte(i.Cb), i.CStride)
+	pCr, err := out.NewPlane(ChannelCr, halfW, halfH, depth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add Cr plane: %v", err)
+	}
+	pCr.SetData([]byte(i.Cr), i.CStride)
+
+	return out, nil
+}
+
+func EncodeFromImage(img image.Image, compression Compression, quality int, lossless losslessMode, logging loggingLevel) (*Context, error) {
+	var out *Image
+
+	switch cm := img.ColorModel(); cm {
+	default:
+		return nil, fmt.Errorf("color model %v not supported yet", cm)
+	case color.NRGBAModel:
+		i, ok := img.(*image.NRGBA)
```

Comment:
Checking the `ColorModel()` is not sufficient to determine the type. There could be implementations of `image.Image` with a color model `color.NRGBAModel` that are not subclasses of `image.NRGBA`.

Better would be to use a type switch (https://tour.golang.org/methods/16).

Reply:
Well, that's why I do the type assertion (you have even commented on this exact line).
Shall I remove the ColorModel() switch and go for a type switch? The current implementation uses both which seems redundant to me, but I have no clue about how images work _at all_ and what needs to be checked.